### PR TITLE
Add Faker.Util module

### PR DESCRIPTION
**Pull request by tobyhinloopen**
Initial proposal for `Faker.Util` as discussed here: https://github.com/igas/faker/issues/84

Feedback, suggestions, discussions, etc. welcome.

I'll update changelog/readme/usage/etc and check the generated docs after feedback

### CODE REVIEW

**Comment by tobyhinloopen** [comment]
Let me kick off by saying I don't like the current `cycle` implementation

File: test/faker/util_test.exs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  test "cycle/2" do
+    list = ~w/a b c/
+    {results, _} = Enum.reduce(1..3, {[], nil}, fn _, {results, cycle_state} ->
+      {result, cycle_state} = Faker.Util.cycle(list, cycle_state)
+      {[result|results], cycle_state}
+    end)
+    assert Enum.sort(results) == list
+  end
```
Comment (tobyhinloopen) [inline test/faker/util_test.exs]: I really want this test to be more readable

File: lib/faker/util.ex [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    - **A**: uppercase alphabet A-Z
+
+  The specifier rules can be overriden using the second argument.
+
+  ## Examples
+
+      format("%2d-%3d %a%A %2d%%") #=> "74-381 sK 32%"
+      format("%8nBATMAN", n: fn() -> "nana " end) #=> "nana nana nana nana nana nana nana nana BATMAN"
+  """
+  def format(format_str, rules \\ [d: &digit/0, A: &upper_alphabet/0, a: &lower_alphabet/0]) do
```
Comment (tobyhinloopen) [inline lib/faker/util.ex]: Missing a spec here

File: lib/faker/util.ex [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    - **a**: lowercase alphabet a-z
+    - **A**: uppercase alphabet A-Z
+
+  The specifier rules can be overriden using the second argument.
+
+  ## Examples
+
+      format("%2d-%3d %a%A %2d%%") #=> "74-381 sK 32%"
+      format("%8nBATMAN", n: fn() -> "nana " end) #=> "nana nana nana nana nana nana nana nana BATMAN"
+  """
```
Comment (tobyhinloopen) [inline lib/faker/util.ex]: Missing a spec here

**Comment by tobyhinloopen** [comment]
@igas Subtle reminder. Feel free to say you don't like it or something :)

File: lib/faker/util.ex [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  def digit do
+    pick @digit
+  end
+
+  @doc """
+  Get a random alphabet character as a string; one of a-z or A-Z
+  """
+  @spec alphabet() :: binary
+  def alphabet do
+    pick [@lowercase_alphabet|@uppercase_alphabet]
```
Comment (igas) [inline lib/faker/util.ex]: maybe we should define `@alphabet` to not concatenate lists at runtime?

File: lib/faker/util.ex [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  end
+
+  @doc """
+  Get a random lowercase character as a string; one of a-z
+  """
+  @spec lower_alphabet() :: binary
+  def lower_alphabet do
```
Comment (igas) [inline lib/faker/util.ex]: alphabet -> letter?

File: lib/faker/util.ex [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      {a, cycle_state} = cycle ~w(1 2 3)
+      {b, cycle_state} = cycle ~w(1 2 3), cycle_state
+      {c, cycle_state} = cycle ~w(1 2 3), cycle_state
+      [a, b, c] #=> [3, 1, 2]
+  """
+  @spec cycle([any], any) :: {any, any}
+  def cycle(_list, _state \\ nil)
+  def cycle(list, nil), do: cycle(list, [])
+  def cycle(list, []), do: cycle(list, Enum.shuffle(list))
+  def cycle(_original_list, [h|t]), do: {h, t}
```
Comment (igas) [inline lib/faker/util.ex]: Maybe we can use agent here? to get PID and grab element from it. Something like:

```
cycle_start([any]) :: pid
cycle(pid) :: any
```

File: lib/faker/util.ex [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  Format a string with randomly generated data. Format specifiers are replaced by random values. A
+  format specifier follows this prototype:
+
+      %[length]specifier
+
+  The following specifier rules are present by default:
+
+    - **d**: digits 0-9
+    - **a**: lowercase alphabet a-z
+    - **A**: uppercase alphabet A-Z
```
Comment (igas) [inline lib/faker/util.ex]: Should we add `n` here?

File: lib/faker/util.ex (hunk 6 above)
Comment (tobyhinloopen) [inline lib/faker/util.ex]: Good idea! I just learned about `Agent` and `GenServer`, I didn't know what they were for before

File: lib/faker/util.ex (hunk 7 above)
Comment (tobyhinloopen) [inline lib/faker/util.ex]: What does `n` do?

Comment (igas) [inline lib/faker/util.ex]: Ah, maybe I misunderstood how it works, I got it from `format("%8nBATMAN", n: fn() -> "nana " end) #=> "nana nana nana nana nana nana nana nana BATMAN"` But I assumed this is predefined. But you already mentioned it in: `The specifier rules can be overriden using the second argument.`

File: lib/faker/util.ex [hunk 8]
```diff
@@ -0,0 +1,127 @@
+defmodule Faker.Util do
+  @digit ~w/0 1 2 3 4 5 6 7 8 9/
+  @lowercase_alphabet ~w/a b c d e f g h i j k l m n o p q r s t u v w x y z/
```
Comment (tobyhinloopen) [inline lib/faker/util.ex]: It might be. What would be locale specific? The alphabet?

Comment (igas) [inline lib/faker/util.ex]: Yeah, like Ru, Fr etc. So all of this functions will return native characters

Comment (igas) [inline lib/faker/util.ex]: Similar to what we do with names for example

**Comment by tobyhinloopen** [comment]
I'll pick up on the feedback in about a week or 2. I'm in the middle of something else

**Comment by igas** [comment]
Hey @tobyhinloopen, any news regarding this one?

**Comment by tobyhinloopen** [comment]
@igas Updated some stuff :)

**Pull request merged**